# scTyper: a comprehensive pipeline for the cell typing analysis of single-cell RNA-seq data
____________________________________________________________________________________________________________________________</br>

## 1. Requirements

### 1.1 Software 

In order to run the entire cell typing pipeline, users need to install the *scTyper* R package from GitHub. See [the installtion instructions](https://github.com/omicsCore/scTyper/blob/master/vignettes/Detailed_installation_instructions.md) for details.

### 1.2 input scRNA-seq data

scTyper uses two types of input data </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- raw scRNA-seq data of 10x genomics platform</br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- pre-processed seurat object : the processing steps of ‘QC’, ‘Cell Ranger’ and ‘Seurat processing’ are done.</br>


## 2. Sample analysis

We reproduce here the analysis detailed in manuscript.

### 2.1 Data Preparation

#### 2.1.1 Single cell data


```r
# load test scRNA-seq datset
system("wget https://github.com/omicsCore/scTyper/blob/master/data/GSE103322.seurat.rda")
test.seurat = get(load(file.path(getwd(), "GSE103322.seurat.rda")))
```

 - Total gene expression data were obtained from [GSE103322](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103322), which generated by single cell sequencing analysis from head and neck squamous cancer of 17 patients, which provided gene expression profiles for 23686 genes in 5,902 cells.
 - This data is already organized and pre-processed (QC, Cell Ranger and Seurat processing have been already processed).
 - lt is available at GitHub **(scTyper/data/GSE103322.seurat.rda)**.

#### 2.1.2 Phenotype data (Required)

```r
# example phenotype data (csv format)
pheno.fn=system.file("extdata/pheno_info_public.csv", package = "scTyper")
```

Users need to make the sample phenotype information table as input for pipeline in a tabular form  **‘.csv’** format. The order of the columns must follow the example template and the names of the columns be formatted as `Sample_ID` and `TissueType`.
We use [Pheotype information](https://github.com/omicsCore/scTyper/raw/master/inst/extdata/pheno_info_public.csv) to testdata. </br>
</br>
 - **Example** </br>
 
| Sample_ID                  | TissueType       | 
|-------------------|-------------|
|  sample_1 | Cancer |
|  sample_2 | Cancer |
|  sample_3 | Normal |
|  sample_4 | Normal |
|  sample_5 | Normal |

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**Sample_ID** : Sample name of single cell RNA sequencing dataset.  </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**TissueType** : Type of tissue that is composed of cells with similar structure and act together to perform a specific function. </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ex) Normal, Cancer etc. </br>


### 2.2 Run scTyper() 
scTyper can be run as a single command  like a following example.
For user’s convenience, the package has been supported with raw data preprocessing pipelines by wrapper functions for FASTQC and Cell Ranger from 10X Genomics; the preprocessing includes quality control, sequence alignment and quantification of raw sequencing data.

 - ***Runing (pre-processing skip, NTP cell typing with inferCNV at cell level)*** </br>
 

```r
celltyped.seurat=scTyper(wd = getwd(),
                         output.name = "cell.typed.seurat.result",
                         pheno.fn = system.file("extdata/pheno_info_public.csv", package = "scTyper"),
                         qc = FALSE,
                         run.cellranger=FALSE,
                         norm.seurat=FALSE,
                         cell.typing.method="NTP",
                         level="cell",
                         run.inferCNV=TRUE,
                         proj.name = "scTyper",
                         seurat.object=test.seurat,
                         marker="Puram.2017.HNSCC.TME",
                         gene.ref.gtf="/data/pubdata/ngs_ref/cellranger/refdata-cellranger-GRCh38-1.2.0/genes/genes.gtf",
                         feature.to.test = "cell.type",
                         cells.test_excluded=c("Epithelial_cell", "Unresolved_cell"),
                         cells.test_reference = c("Fibroblasts", "T.cells", "B_Plasma.cells", "Macrophages", "Dendritic.cells", "Endothelial.cells", "Myocytes", "Mast"),
                         malignant.cell.type="Epithelial_cell",
                         report.mode=TRUE,
                         mc.cores = 10)
```


### 2.3 scTyper() arguments description

scTyper is an R package which can be executed by a single command. In the table below, the user can check the argument and description of scTyper().
For a detailed parameter description of the single command and its internal step-by-step functions, see the [Reference Manual](https://github.com/omicsCore/scTyper/files/4763300/scTyper_reference_manual.pdf).

<table class="table table-striped" style="width: auto !important; ">
 <thead>
  <tr>
   <th style="text-align:left;"> Process </th>
   <th style="text-align:left;"> Parameters </th>
   <th style="text-align:left;"> Description </th>
   <th style="text-align:left;"> Values </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> global configuration </td>
   <td style="text-align:left;"> wd </td>
   <td style="text-align:left;"> Working directory </td>
   <td style="text-align:left;"> Character </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> output.name </td>
   <td style="text-align:left;"> Output directory name </td>
   <td style="text-align:left;"> Character </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> pheno.fn </td>
   <td style="text-align:left;"> Phenotype file path </td>
   <td style="text-align:left;"> File path </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> qc, run.cellranger , norm.seurat </td>
   <td style="text-align:left;"> Indicate whether the process run </td>
   <td style="text-align:left;"> Logical (Default = ‘FALSE’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> cell.typing.method </td>
   <td style="text-align:left;"> Cell typing method </td>
   <td style="text-align:left;"> ‘NTP’ (Default), ‘ES’, ‘Average’ </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> level </td>
   <td style="text-align:left;"> Indicate the cell assignment level (cell or cluster) </td>
   <td style="text-align:left;"> ‘cell’ (Default), ‘cluster’ </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> run.inferCNV </td>
   <td style="text-align:left;"> Indicate whether ‘malignant cell typing by inferCNV process run </td>
   <td style="text-align:left;"> Logical (Default = ‘TRUE’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> project.name </td>
   <td style="text-align:left;"> Project name </td>
   <td style="text-align:left;"> Character </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> mc.cores </td>
   <td style="text-align:left;"> Number of cores </td>
   <td style="text-align:left;"> Numeric (Default = ‘1’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> report.mode </td>
   <td style="text-align:left;"> Generate report file </td>
   <td style="text-align:left;"> Logical (Default = ‘TRUE’) </td>
  </tr>
  <tr>
   <td style="text-align:left;"> QC </td>
   <td style="text-align:left;"> fastqc.path </td>
   <td style="text-align:left;"> FastQC program path </td>
   <td style="text-align:left;"> File path </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> fastq.dir </td>
   <td style="text-align:left;"> FastQC output directory </td>
   <td style="text-align:left;"> File path </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> fq1.idx </td>
   <td style="text-align:left;"> Index of the FASTQ file (Read 1) </td>
   <td style="text-align:left;"> Character (Default = ‘_R1_001.fastq’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> fq2.idx </td>
   <td style="text-align:left;"> Index of the FASTQ file (Read 2) </td>
   <td style="text-align:left;"> Character (Default = ‘_R2_001.fastq’) </td>
  </tr>
  <tr>
   <td style="text-align:left;"> CellRanger </td>
   <td style="text-align:left;"> cellranger.path </td>
   <td style="text-align:left;"> Cell Ranger program path </td>
   <td style="text-align:left;"> File path </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> cellranger.ref.dir </td>
   <td style="text-align:left;"> Directory of Cell Ranger reference file </td>
   <td style="text-align:left;"> File path </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Seurat processing </td>
   <td style="text-align:left;"> percent.min.cells </td>
   <td style="text-align:left;"> Cutoff to filter features containing minimum percent of cells </td>
   <td style="text-align:left;"> 0.1 </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> min.features </td>
   <td style="text-align:left;"> Cutoff to filter cells containing minimum number of features </td>
   <td style="text-align:left;"> 200 </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> percent.mt </td>
   <td style="text-align:left;"> Cutoff for filtering cells that have &gt;n percent mitochondrial counts </td>
   <td style="text-align:left;"> 10 </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> vars.to.regress </td>
   <td style="text-align:left;"> Variables to regress out </td>
   <td style="text-align:left;"> Default=c(‘nCount_RNA’, ‘percent.mt’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> dims </td>
   <td style="text-align:left;"> A vector of the dimensions to use in construction of the SNN graph. </td>
   <td style="text-align:left;"> 1:100 </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> resolution </td>
   <td style="text-align:left;"> Value of the resolution parameter, use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities. </td>
   <td style="text-align:left;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Cell typing </td>
   <td style="text-align:left;"> seurat.object </td>
   <td style="text-align:left;"> Seurat object </td>
   <td style="text-align:left;"> Seurat object </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> slot </td>
   <td style="text-align:left;"> Data type of Seurat object </td>
   <td style="text-align:left;"> ‘scale.data’ (Default), ‘count.data’, ‘data’ </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> marker </td>
   <td style="text-align:left;"> Cell markers to use cell typing </td>
   <td style="text-align:left;"> Character or List (Signature names or Study names or User defined gene set list) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> assay </td>
   <td style="text-align:left;"> Assay of Seurat object </td>
   <td style="text-align:left;"> Character (Default=’RNA’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> NTP.g.filter.method </td>
   <td style="text-align:left;"> Method to filter genes in NTP </td>
   <td style="text-align:left;"> ‘sd’ (Default),’mad’, ‘none’ </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> NTP.gene.filter.cutoff </td>
   <td style="text-align:left;"> Cutoff to filter genes of in NTP </td>
   <td style="text-align:left;"> Numeric (Default = ‘0.3’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> NTP.distance </td>
   <td style="text-align:left;"> NTP distance method </td>
   <td style="text-align:left;"> ‘cosine’ (Default), ‘correlation’ </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> NTP.norm.method </td>
   <td style="text-align:left;"> NTP normalization method </td>
   <td style="text-align:left;"> ‘none’ (Default), ‘row.std’ </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Malignant cell typing (inferCNV </td>
   <td style="text-align:left;"> gene.ref.gtf </td>
   <td style="text-align:left;"> Path of GTF file including genomic location for genes </td>
   <td style="text-align:left;"> File path </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> feature.to.test </td>
   <td style="text-align:left;"> Column header name of the meta data in Seurat object (select the cell groups for T.test) </td>
   <td style="text-align:left;"> Character (Default = ‘cell.type’), ‘tissue.type’ </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> cells.test_excluded </td>
   <td style="text-align:left;"> A value indicates the cells to be excluded in T.test </td>
   <td style="text-align:left;"> Character (Default = ‘Epithelial’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> cells.test_reference </td>
   <td style="text-align:left;"> A value indicates the cells to use as be excluded in T.test </td>
   <td style="text-align:left;"> Character (Default = ‘immune’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> fc.cutoff </td>
   <td style="text-align:left;"> Cutoff of fold change </td>
   <td style="text-align:left;"> Numeric (Default = ‘0.05’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> cutoff.gene.cluster </td>
   <td style="text-align:left;"> A cutoff P-value for filtering out the gene clusters (calculated from GO analysis) </td>
   <td style="text-align:left;"> Numeric (Default = ‘0.05’) </td>
  </tr>
  <tr>
   <td style="text-align:left;">  </td>
   <td style="text-align:left;"> malignant.cell.type </td>
   <td style="text-align:left;"> Cell type to assign malignant cell </td>
   <td style="text-align:left;"> Character (Default = ’Epithelial’) </td>
  </tr>
</tbody>
</table>
</br>
  - **Warning**: When users have pre-processed seurat object (completed with qc, Cell Ranger and Seurat processing), **seurat.object** parameter must be assigned.
 
#### 2.3.1 **'marker'** parameter insertion of scTyper

The scTyper package provides manually curated and pre-installed cell marker databases comprised of 'scTyper.db' and 'CellMarker DB' for cell typing. Users can choose the cell markers of interest from [marker database](https://github.com/omicsCore/scTyper/files/4723417/cell.marker.database.xlsx).
 
 - **Example** </br>

```r
  # StudyName
  marker="Puram.2017.HNSCC.TME" 
  marker=c("Kawai.2018.Liver", "Li.2017.CRC.TME") 
    
  # Identifier
  marker=c("Costea.2013.OSCC.CAF:Normal cell:NF", "Costea.2013.OSCC.CAF:Cancer cell:CAF_D", 
              "Elyada.2019.PDAC.CAF:Cancer cell:iCAF", "Elyada.2019.PDAC.CAF:Cancer cell:myCAF")
    
  # user defined marker list             
  marker=list(T_cell=c("CD2", "CD3D", "CD3E", "CD3G"),
              Fibroblast=c("FAP", "PDPN", "COL1A2", "DCN", "COL3A1", "COL6A1"),
              Macrophage=c("CD14", "CD163", "CD68", "FCGR2A", "CSF1R"),
              Dendritic_cell=c("CD40", "CD80", "CD83", "CCR7"),
              Mast_cell=c("CMA1", "MS4A2", "TPSAB1", "TPSB2"),
              Myocyte=c("ACTA1", "ACTN2", "MYL2", "MYH2"),
              Endothelial.cell=c("PECAM1", "VWF", "ENG"),
              B_Plasma_cell=c("SLAMF7", "CD79A", "BLNK", "FCRL5"),
              Epithelial_cell=c("KRT14", "KRT17", "KRT6A", "KRT5", "KRT19", "KRT8",
                                "KRT16", "KRT18", "KRT6B", "KRT15", "KRT6C", "KRTCAP3","EPCAM", "SFN"))
```

- User can easily search the markers of interest from these databases, which are uniquely labeled using unified nomenclature. For example, a cell marker label ‘Puram.2017.HNSCC.TME’ was designated by concatenating the first author name of the publication (‘Puram’), publication year (‘2017’), tissue type/cancer type (‘HNSCC’), and categories of cell composition (‘TME’, tumor microenvironment).
- The input of marker parameter is elements of `#Identifier`, `#StudyName` columns or user can define marker list as input parameter
- **Note**: If there is a duplicate gene in the selected marker, it is excluded from the cell typing process.
- **Note**: If there is only one gene in the cell type of the marker, the user should use "Average" of the cell typing method.
- **Update of marker database** </br>
After changing the'sigTyper.db.txt' file in the'extdata' directory to the latest version, you can use the "update.sig.db()" function to update the cell marker database.
    
```r
# example
update.sig.db(sig.db.path = system.file(sig.db.path="extdata/sigTyper.DB.txt", package = "scTyper"), db.name=c("sigTyper.db"), output.dir=system.file("/data",package = "scTyper"))
```

</br>

#### 2.3.2 **inferCNV** parameter description

When `run.inferCNV=TRUE`, users should be careful with confusing parameter insertion. </br>
**feature.to.test** :	Cell group for t.test in the inferCNV process. The value is "tissue.type" or "cell.type", the column header name of the metadata in the Seurat object. </br>
**cells.test_excluded** The value is a cell to be excluded from the inferCNV process. </br>
**cells.test_reference** The value containing the classification of the reference cell to be used by the cell to infer cnv in T.test </br>


### 2.4 scTyper() final output 

 - ***final output of scTyper() is a Seurat object*** : scTyper() function returns a **Seurat object**, which is a popular data type for the subsequent analyses and biological interpretation of scRNA-seq. The seurat object includes the inferred cell type and cnv core as a result of cell typing. The Seurat object is a class allowing for the storage and manipulation of single-cell data. Seurat object was designed to allow for greater flexibility to work with all these data types in a cohesive framework. 

    - Analysis result column name: </br>
        cell.type: Inferred cell type result </br>
        cnv.score: copy number variation score estimated by malignant cell typing (inferCNV) </br>
        malignant.st: logical value of malignant cell estimated by malignant cell typing (inferCNV) </br>

    
```r
# The returned object of the scTyper() function
celltyped.seurat
```
    
```
An object of class Seurat 
23686 features across 5902 samples within 1 assay 
Active assay: RNA (23686 features, 2000 variable features)
2 dimensional reductions calculated: pca, tsne
```


 - ***Report file*** : If `report.mode=TRUE` in scTyper() parameter, the results and the executed processes are automatically documented as a report. `report.mode` is a function used to combine the results into one unified file. The report summarizes the processing steps, cell typing and clustering results, and visualizing the results plots. This may help users reproduce their analysis workflows.
 
 - ***Automatically generated output directory*** : The result files are created automatically in output directory that users setted by parameter ‘wd’ and ‘output.name’ according to your processing step.</br>
 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- 00_qc : Directory of FastQC output </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- 01_count : Directory of CellRanger output </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- 02_NTP : Directory of NTP output </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- 03_inferCNV : Directory of inferCNV output </br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- data : Directory of RData (rda). 'seurat.rda' (final output) saved. </br>


### 2.5 Visualization

If ***“report.mode=TRUE”*** is set true in the scTyper(), the user will get a seurat object as an output.

- **Cell type statistics across samples** </br>


```r
t.tbl=table(celltyped.seurat$cell.type)
tbl=cbind(t.tbl, table(celltyped.seurat$cell.type, celltyped.seurat$sample.name))
colnames(tbl)[1]="Total"
panderOptions('table.split.table', Inf)
pander(tbl, split.cells = 10)
```


-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
        &nbsp;          Total   Cancer_P0   Cancer_P10   Cancer_P12   Cancer_P13   Cancer_P16   Cancer_P17   Cancer_P18   Cancer_P20   Cancer_P22   Cancer_P23   Cancer_P24   Cancer_P25   Cancer_P26   Cancer_P28   Cancer_P5   Cancer_P6   Cancer_P7   Cancer_P8   LymphNode_P0   LymphNode_P20   LymphNode_P25   LymphNode_P26   LymphNode_P28   LymphNode_P5 
---------------------- ------- ----------- ------------ ------------ ------------ ------------ ------------ ------------ ------------ ------------ ------------ ------------ ------------ ------------ ------------ ----------- ----------- ----------- ----------- -------------- --------------- --------------- --------------- --------------- --------------
  **B_Plasma_cell**      190        5           44           2            10           12           6            18           0            0            0            0            6            0            0            3           4           0          13            12              0              27               3              22              3       

  **Dendritic_cell**     222       17           4            12           35           24           8            30           0            5            0            1            30           2            8            5           1           0          10            1               0              19               4               4              2       

 **Endothelial_cell**    250        2           7            0            2            36           18           18           2            1            0            2            5            3            17           9          42           0          26            1               0               9              10              12              28      

    **Fibroblast**       637        8           23           0            2            57           38           29           7            26           0            55           48           33          104          15          39           1          19            4               0              31              38              52              8       

    **Macrophage**       125        4           1            1            4            13           2            23           1            0            0            4            5            2            8           41           3           0           3            0               0               4               1               4              1       

  **Malignant_cell**    2739       20           5            12           16          102          357          214          329          123           0            30          191           63           78          108         211          6          12            0              343             91              217             112             99      

    **Mast_cell**        125        6           1            0            4            25           7            12           0            22           0            20           11           1            0            4           1           0           3            1               0               3               0               2              2       

     **Myocyte**         51         2           0            0            0            4            0            2            0            0            0            0            15           1            16           1           0           0           0            0               0               5               1               4              0       

      **T_cell**        1008       184          0           121           10          269           45          161           0            0            50           0            66           3            0           23           0           0          18            1               0              56               1               0              0       

 **Unresolved_cell**     555       11           4            9            7            54           9            60           6            24           1            18           57           22          139           4           7           0           0            3               3              46              23              34              14      
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

- ***Make table of inferred cell types*** </br>


```r
tbl=table(celltyped.seurat$cell.type)
s.tbl=table(celltyped.seurat$cell.type, celltyped.seurat$sample.name); r.tbl=apply(s.tbl, 2, function(a) a/sum(a))
```

- **Distribution of inferred cell types** </br>


```r
par(mar=c(7,5,0,0));barplot(tbl, las=2, ylab="Number of cells", cex.names=0.8, cex.axis=0.8)
```

![](Sample_analysis_files/figure-html/unnamed-chunk-10-1.png)<!-- -->

- **Distibution of cell types across samples** </br>


```r
par(mar=c(7,5,0,10))
barplot(r.tbl, las=2, ylab="Proportion (%)", cex.names=0.8, cex.axis=0.8, legend.text = rownames(r.tbl), args.legend = list(x = 'right', bty='n', inset=c(-0.3,0), xpd = TRUE, cex=0.7))
```

![](Sample_analysis_files/figure-html/unnamed-chunk-11-1.png)<!-- -->


- **inferred cell types by cell typing method** </br>


```r
cols=rainbow_hcl(length(levels(celltyped.seurat$cell.type))); names(cols)=levels(celltyped.seurat$cell.type)
p1=DimPlot(celltyped.seurat, reduction = 'tsne',group.by="cell.type",label=F,cols=cols, pt.size=0.2); p1=LabelClusters(plot = p1, id = "cell.type", size = 3)
p1
```

![](Sample_analysis_files/figure-html/unnamed-chunk-12-1.png)<!-- -->

- **Malignant cells by inferCNV** </br>


```r
p4=DimPlot(celltyped.seurat, reduction = 'tsne',group.by="malignant.st",label=F, pt.size=0.2, cols=c("grey90", "red")) 
p4
```

![](Sample_analysis_files/figure-html/unnamed-chunk-13-1.png)<!-- -->

- **CNV score by inferCNV** </br>

![](Sample_analysis_files/figure-html/unnamed-chunk-14-1.png)<!-- -->

- **Seurat clusters** </br>


```r
p2=DimPlot(celltyped.seurat, reduction = 'tsne',group.by="seurat_clusters",label=F, pt.size=0.2); p2=LabelClusters(plot = p2, id = "seurat_clusters", size = 3)
p2
```

![](Sample_analysis_files/figure-html/unnamed-chunk-15-1.png)<!-- -->

- **Samples** </br>


```r
p3=DimPlot(celltyped.seurat, reduction = 'tsne',group.by="sample.name",label=F, pt.size=0.2)
p3
```

![](Sample_analysis_files/figure-html/unnamed-chunk-16-1.png)<!-- -->

- **Cell markers heatmap** </br>
A heatmap shows the cell typing result and the gene expression levels of cell marker gene sets from Puram.2017.HNSCC.TME. For each method, the assigned cell types are indicated by color bars.  


```r
draw.heatmap(seurat = celltyped.seurat,
             wd = "/data/Rpackage/scTyper",
             run.inferCNV = TRUE,
             slot = "scale.data",
             marker="Puram.2017.HNSCC.TME")
```

![](Sample_analysis_files/figure-html/unnamed-chunk-17-1.png)<!-- -->



### 2.6 Example script for different customized pipeline

 - ***Runing (pre-processing skip, NTP cell typing with inferCNV at cell level)***
 

```r
celltyped.seurat=scTyper(wd = "/data/Rpackage/scTyper",
                         output.name = "cell.typed.seurat.result",
                         pheno.fn = "/data/Rpackage/scTyper/data/pheno_info_public.csv",
                         qc = FALSE,
                         run.cellranger=FALSE,
                         norm.seurat=FALSE,
                         cell.typing.method="NTP",
                         level="cell",
                         run.inferCNV=TRUE,
                         proj.name = "scTyper",
                         seurat.object=test.seurat,
                         marker="Puram.2017.HNSCC.TME",
                         gene.ref.gtf="/data/pubdata/ngs_ref/cellranger/refdata-cellranger-GRCh38-1.2.0/genes/genes.gtf",
                         feature.to.test = "cell.type",
                         cells.test_excluded=c("Epithelial_cell", "Unresolved_cell"),
                         cells.test_reference = c("Fibroblasts", "T.cells", "B_Plasma.cells", "Macrophages", "Dendritic.cells", "Endothelial.cells", "Myocytes", "Mast"),
                         malignant.cell.type="Epithelial_cell",
                         report.mode=TRUE,
                         mc.cores = 10)
```

 - ***Runing (pre-processing skip, ES cell typing with inferCNV at cell level)***
 

```r
celltyped.seurat=scTyper(wd = "/data/Rpackage/scTyper",
                         output.name = "cell.typed.seurat.result",
                         pheno.fn = "/data/Rpackage/scTyper/data/pheno_info_public.csv",
                         qc = FALSE,
                         run.cellranger=FALSE,
                         norm.seurat=FALSE,
                         cell.typing.method="ES",
                         level="cell",
                         run.inferCNV=TRUE,
                         proj.name = "scTyper",
                         seurat.object=test.seurat,
                         marker="Puram.2017.HNSCC.TME",
                         gene.ref.gtf="/data/pubdata/ngs_ref/cellranger/refdata-cellranger-GRCh38-1.2.0/genes/genes.gtf",
                         feature.to.test = "cell.type",
                         cells.test_excluded=c("Epithelial_cell", "Unresolved_cell"),
                         cells.test_reference = c("Fibroblasts", "T.cells", "B_Plasma.cells", "Macrophages", "Dendritic.cells", "Endothelial.cells", "Myocytes", "Mast"),
                         malignant.cell.type="Epithelial_cell",
                         report.mode=TRUE,
                         mc.cores = 10)
```
 
 - ***Runing (pre-processing skip, Average cell typing with inferCNV at cell level)***
 

```r
celltyped.seurat=scTyper(wd = "/data/Rpackage/scTyper",
                         output.name = "cell.typed.seurat.result",
                         pheno.fn = "/data/Rpackage/scTyper/data/pheno_info_public.csv",
                         qc = FALSE,
                         run.cellranger=FALSE,
                         norm.seurat=FALSE,
                         cell.typing.method="Average",
                         level="cell",
                         run.inferCNV=TRUE,
                         proj.name = "scTyper",
                         seurat.object=test.seurat,
                         marker="Puram.2017.HNSCC.TME",
                         gene.ref.gtf="/data/pubdata/ngs_ref/cellranger/refdata-cellranger-GRCh38-1.2.0/genes/genes.gtf",
                         feature.to.test = "cell.type",
                         cells.test_excluded=c("Epithelial_cell", "Unresolved_cell"),
                         cells.test_reference = c("Fibroblasts", "T.cells", "B_Plasma.cells", "Macrophages", "Dendritic.cells", "Endothelial.cells", "Myocytes", "Mast"),
                         malignant.cell.type="Epithelial_cell",
                         report.mode=TRUE,
                         mc.cores = 10)
```
 
 - ***Runing (pre-processing(QC, CellRanger, Normalization) , NTP cell typing with inferCNV at cell level)***
 

```r
###### Running all step (User's data)
processed.celltyped.seurat=scTyper(wd="/data/Rpackage/scTyper",
                                   output.name = "test.result",
                                   pheno.fn="/data/Rpackage/scTyper/data/pheno_info_test.csv",
                                   qc = TRUE,
                                   run.cellranger=TRUE,
                                   norm.seurat=TRUE,
                                   cell.typing.method="NTP",
                                   level="cell",
                                   run.inferCNV=TRUE,
                                   proj.name = "scTyper",
                                   fastqc.path="/data/program/bin/fastqc",
                                   fastq.dir="/data/Rpackage/scTyper/test",
                                   fq1.idx="_R1_001.fastq",
                                   fq2.idx="_R2_001.fastq",
                                   cellranger.path="/data/program/bin/cellranger",
                                   cellranger.ref.dir="/data/pubdata/ngs_ref/cellranger/refdata-cellranger-GRCh38-1.2.0",
                                   marker="Puram.2017.HNSCC.TME",
                                   gene.ref.gtf="/data/pubdata/ngs_ref/cellranger/refdata-cellranger-GRCh38-1.2.0/genes/genes.gtf",
                                   feature.to.test = "tissue.type",
                                   cells.test_excluded=c("Epithelial_cell", "Unresolved_cell"),
                                   cells.test_reference = "Normal",
                                   malignant.cell.type="Epithelial_cell",
                                   report.mode=TRUE,
                                   mc.cores = 10)
```

Help information is available with `scTyper --help`.
<br>
<br>

## 3. Reference

1.	Hwang B, Lee JH, Bang D: Single-cell RNA sequencing technologies and bioinformatics pipelines. Experimental & molecular medicine 2018, 50(8):96.
2.	Abdelaal T, Michielsen L, Cats D, Hoogduin D, Mei H, Reinders MJT, Mahfouz A: A comparison of automatic cell identification methods for single-cell RNA sequencing data. Genome Biol 2019, 20(1):194-194.
3.	Pliner HA, Shendure J: Supervised classification enables rapid annotation of cell atlases. 2019, 16(10):983-986.
4.	Ma F, Pellegrini M: ACTINN: automated identification of cell types in single cell RNA sequencing. Bioinformatics (Oxford, England) 2020, 36(2):533-538.
5.	Alquicira-Hernandez J, Sathe A, Ji HP, Nguyen Q, Powell JE: scPred: accurate supervised method for cell-type classification from single-cell RNA-seq data. Genome Biol 2019, 20(1):264.
6.	Kim T, Lo K, Geddes TA, Kim HJ, Yang JYH, Yang P: scReClassify: post hoc cell type classification of single-cell rNA-seq data. 2019, 20(Suppl 9):913.
7.	Ceder JA, Jansson L, Helczynski L, Abrahamsson PA: Delta-like 1 (Dlk-1), a novel marker of prostate basal and candidate epithelial stem cells, is downregulated by notch signalling in intermediate/transit amplifying cells of the human prostate. European urology 2008, 54(6):1344-1353.
8.	Ma S, Chan KW, Hu L, Lee TK, Wo JY, Ng IO, Zheng BJ, Guan XY: Identification and characterization of tumorigenic liver cancer stem/progenitor cells. Gastroenterology 2007, 132(7):2542-2556.
9.	Zhang X, Lan Y, Xu J, Quan F, Zhao E, Deng C, Luo T, Xu L, Liao G, Yan M et al: CellMarker: a manually curated resource of cell markers in human and mouse. Nucleic acids research 2019, 47(D1):D721-d728.
10.	Costea DE, Hills A, Osman AH, Thurlow J, Kalna G, Huang X, Pena Murillo C, Parajuli H, Suliman S, Kulasekara KK et al: Identification of two distinct carcinoma-associated fibroblast subtypes with differential tumor-promoting abilities in oral squamous cell carcinoma. Cancer research 2013, 73(13):3888-3901.
11.	Navab R, Strumpf D, Bandarchi B, Zhu CQ, Pintilie M, Ramnarine VR, Ibrahimov E, Radulovich N, Leung L, Barczyk M et al: Prognostic gene-expression signature of carcinoma-associated fibroblasts in non-small cell lung cancer. Proceedings of the National Academy of Sciences of the United States of America 2011, 108(17):7160-7165.
12.	Zhang Q, He Y, Luo N, Patel SJ, Han Y, Gao R, Modak M, Carotta S, Haslinger C, Kind D et al: Landscape and Dynamics of Single Immune Cells in Hepatocellular Carcinoma. Cell 2019, 179(4):829-845.e820.
13.	Zheng GXY, Terry JM, Belgrader P, Ryvkin P, Bent ZW, Wilson R, Ziraldo SB, Wheeler TD, McDermott GP, Zhu J et al: Massively parallel digital transcriptional profiling of single cells. Nature Communications 2017, 8(1):14049.
14.	Satija R, Farrell JA, Gennert D, Schier AF, Regev A: Spatial reconstruction of single-cell gene expression data. Nature biotechnology 2015, 33(5):495-502.
15.	Hoshida Y: Nearest template prediction: a single-sample-based flexible class prediction with confidence assessment. PloS one 2010, 5(11):e15543.
16.	Subramanian A, Tamayo P, Mootha VK, Mukherjee S, Ebert BL, Gillette MA, Paulovich A, Pomeroy SL, Golub TR, Lander ES et al: Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences of the United States of America 2005, 102(43):15545-15550.
17.	Patel AP, Tirosh I, Trombetta JJ, Shalek AK, Gillespie SM, Wakimoto H, Cahill DP, Nahed BV, Curry WT, Martuza RL et al: Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma. Science (New York, NY) 2014, 344(6190):1396-1401.
18.	Puram S, Tirosh I, Parikh A, Patel A, Yizhak K, Gillespie S, Rodman C, Luo C, Mroz E, Emerick K et al: Single-Cell Transcriptomic Analysis of Primary and Metastatic Tumor Ecosystems in Head and Neck Cancer. Cell 2017, 171.




